### ISSUE TITLE
IDocumentWriter - support writing to response stream directly?

### ISSUE THREAD

**Issue by DSilence** [graphql-dotnet/graphql-dotnet#798]
## Summary

The current document writer only support returning a serialized string, which significantly limits the supported serializer usage. Fast json serializers for .net, like [utf8json](https://github.com/neuecc/Utf8Json) or  [SpanJson](https://github.com/Tornhoof/SpanJson) support writing data directly to stream asynchronously, which positively impacts performance, since they don't allocate a string. In addition, such approach would allow the usage of binary protocols in future.

### Proposal

Change the IDocumentWriter interface to look similar to this
```
public interface IDocumentWriter
  {
    Task WriteAsync(Stream stream, object value);
  }
```

or, even better, use generic method

```
public interface IDocumentWriter
  {
    Task WriteAsync<T>(Stream stream, T value);
  }
```
### Motivation

Better performance. This is a breaking change.

**Comment by DSilence** [comment]
I'm willing to work on a PR if you guys are ok with it.

**Comment by joemcbride** [comment]
Yes ... here's a thought.  `DocumentWriter` is meant to be used with `ExecutionResult`.  I can't think of any other use for it.  Perhaps we should just make it specific to that.  It would also be good to provide an upgrade path with marking the previous method `Obsolete`, which we would then remove in the next major version.

```csharp
public interface IDocumentWriter
{
  Task WriteAsync(Stream stream, ExecutionResult value);

  [Obsolete("This method is obsolete and will be removed in the next major version.  Use WriteAsync instead.")]
  string Write(object value);
}
```

**Comment by DSilence** [comment]
I've given some thought to the WebSocket transport. The following solution might work:

1. IDocumentWriter should be able to write OperationMessage. This could be achieved using the original generic Write method concept (possibly with some generic restriction only allowing ExecutionResult and OperationMessage).
2. The OperationMessage payload type should be changed to object or generic. This way we can store ExecutionResult without turning it into JObject for writing purposes. 
3. -  Option A: In WebSocketWriterPipeline we create a MemoryStream and use IDocumentWriter to write OperationMessage to the stream. We can then use TryGetBuffer to get the ArraySegment and send it over socket.  (Optional) The MemoryStream instances used by WebSocketWriterPipeline could use an ArrayPool to avoid extra allocations. 
    -  Option B: We add an additional method to IDocumentWriter to write the object to ArraySegment or Memory. We can then use the produced entity to write it to a WebSocket channel.

Alternative: A separate entity could be created specifically for writing OperationMessage.

**Comment by DSilence** [comment]
Ping @BenjaBobs, @joemcbride. What do you think of this proposal?

**Comment by BenjaBobs** [comment]
I like the idea of option 2.

Something like: 
```c#
public class OperationMessage
{
    public string Id { get; set; }
    public string Type { get; set; }
    public object Payload { get; set; }
}

public class OperationMessage<TPayload> : OperationMessage
{
    public new TPayload Payload 
    { 
        get => (TPayload)base.payload; 
        set => base.payload = value; 
    }
}
```

And then we can use the IDocumentWriter to serialize it in the WebSocketWriterPipeline.

https://github.com/graphql-dotnet/server/blob/develop/src/Transports.Subscriptions.WebSockets/WebSocketWriterPipeline.cs#L51-L82

**Comment by DSilence** [comment]
Sounds good. Here is what I've had in mind for IDocumentWriter for that case. The challenge here is that to pool the byte array to avoid unnecessary allocations. So we need a proxy class for result that will be able to store the results of the serialization and return the array pool back after the response was processed.
The interface: 
```csharp
public interface IByteResult : IDisposable
    {
        ArraySegment<byte> Result { get; }
    }
```
The implementation (note that a reference to System.Buffers is required):
```csharp
public class PooledByteResult : IByteResult
    {
        private readonly byte[] _buffer;
        private readonly ArrayPool<byte> _pool;

        public PooledByteResult(ArrayPool<byte> pool, int minLength)
        {
            _pool = pool;
            _buffer = _pool.Rent(minLength);
            Stream = new MemoryStream(_buffer);
        }

        public ArraySegment<byte> Result { get; private set; }

        internal MemoryStream Stream { get; }

        internal void InitResponseFromCurrentStreamPosition()
        {
            Result = new ArraySegment<byte>(_buffer, 0, (int)Stream.Position);
        }

        private void ReleaseUnmanagedResources()
        {
            _pool.Return(_buffer);
            Stream.Dispose();
        }

        public void Dispose()
        {
            ReleaseUnmanagedResources();
            GC.SuppressFinalize(this);
        }

        ~PooledByteResult()
        {
            ReleaseUnmanagedResources();
        }
    }
```
Resulting API for IDocumentWriter will look like this:
```csharp
public interface IDocumentWriter
    {
        Task WriteAsync<T>(Stream stream, T value);

        Task<IByteResult> WriteAsync<T>(T value);

        [Obsolete("This method is obsolete and will be removed in the next major version.  Use WriteAsync instead.")]
        string Write(object value);
    }
```
Implementation:
```csharp
public class DocumentWriter : IDocumentWriter
    {
        private readonly ArrayPool<byte> _pool = ArrayPool<byte>.Shared;
        private readonly int _maxArrayLength = 1048576;
        private readonly JsonSerializer _serializer;
        private static readonly Encoding Utf8Encoding = new UTF8Encoding(false);

        public DocumentWriter()
            : this(indent: false)
        {
        }

        public DocumentWriter(bool indent)
            : this(
                indent ? Formatting.Indented : Formatting.None,
                new JsonSerializerSettings())
        {
        }

        public DocumentWriter(Formatting formatting, JsonSerializerSettings settings)
        {
            _serializer = JsonSerializer.CreateDefault(settings);
            _serializer.Formatting = formatting;
        }

        public Task WriteAsync<T>(Stream stream, T value)
        {
            using (var writer = new StreamWriter(stream, Utf8Encoding, 1024, true))
            using (var jsonWriter = new JsonTextWriter(writer))
            {
                _serializer.Serialize(jsonWriter, value);
            }

            return TaskExtensions.CompletedTask;
        }

        public async Task<IByteResult> WriteAsync<T>(T value)
        {
            var pooledDocumentResult = new PooledByteResult(_pool, _maxArrayLength);
            var stream = pooledDocumentResult.Stream;
            try
            {
                await WriteAsync(stream, value).ConfigureAwait(false);
                pooledDocumentResult.InitResponseFromCurrentStreamPosition();
                return pooledDocumentResult;
            }
            catch (Exception)
            {
                pooledDocumentResult.Dispose();
                throw;
            }
        }

        public string Write(object value)
        {
            return this.WriteToStringAsync((ExecutionResult) value).GetAwaiter().GetResult();
        }
    }
```

I believe this covers all our cases.

**Comment by joemcbride** [comment]
Merged in #809
